Bluetooth: Controller: Rx node piggy v2

erbr-ot opened the pull request:
Fixes #45323.

When possible re-use the already allocated RX node for notifications.
Store (retain) RX node and Link element on RX if NTF could occur.
Pass link element to LLCP (ull_cp_rx()) together with RX node.
New RX node type RETAIN introduced to signal retention

When no RX node is available allocate one and hold off TX on procedures
until such time that a node is available for NTF.

In case waiting for NTF buffer avail is needed, allocate and store TX
node to use for TX once NTF becomes available.

On termination of connection, pass list of retained rx & tx nodes (if any) to conn cleanup
for recycling.
 
CIS Established (incl. timeout handling) is now handled entirely as a
specific event driven by ull_conn_iso - ie removal of procedure check
of cis->established and cis->expire, as this is doubling mechanism
in the conn_iso context.

Unit test and helpers updated to handle new node type.
Function ull_cp_release_ntf() was used only in unit test,
so moved to helper context.
Updating release_ntf to handle the fact that with piggy-backing in test
context the node used for NTF can be from two different memory pools

Basic connection Bsim test extended to perform multiples of re-connect to same peripheral

erbr-ot committed:
Bluetooth: controller: use RX node piggy-back for NTF when possible

When possible re-use the already allocated RX node for notifications.
Store (retain) RX node and Link element on RX if NTF could occur.
Pass link element to LLCP (ull_cp_rx()) together with RX node.
New RX node type RETAIN introduced to signal retention

When no RX node is available allocate one and hold off TX on procedures
until such time that a node is available for NTF.

In case waiting for NTF buffer avail is needed, allocate and store TX
node to use for TX once NTF becomes available.

CIS Established (incl. timeout handling) is now handled entirely as a
specific event driven by ull_conn_iso - ie removal of procedure check
of cis->established and cis->expire, as this is doubling mechanism
in the conn_iso context.

Unit test and helpers updated to handle new node type.
Function ull_cp_release_ntf() was used only in unit test, so moved to
helper context.
Updating release_ntf to handle the fact that with piggy-backing in test
context the node used for NTF can be from two different memory pools

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
Bluetooth: controller: fixing CIS establish flow for refactored LLCP

Latest modification to LLCP removed 'monitoring' of cis->established
flag in LLCP, relying instead on signal through ULL.

For this to work properly the signal must also be generated, so this is
introduced. Note - checking of cis->established flag could be removed
if caller ensures only calling once (on establish condition)

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
Bluetooth: controller: updating TX node pre-alloc store mechanism

Instead of using the node ref used for TX ack for storing a TX node
pre-allocated in case of NTF RX node availability delay, introduce a
separate store to TX node. This to avoid misunderstanding on clean-up
in terminate flow.

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
Bluetooth: controller: adding release of rx/tx nodes retained in LLCP

On connection clean-up release nodes held by LLCP. These are released
in two steps, first on LLCP terminate (in local/remote) by moving
to rx/tx_node_release chains, and then on conn_cleanup_finalize
nodes are released back to appropriate mem-pool/queue.
Rename abort functions to refer to terminate.

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

sjanc commented:
#AutoPTS run zephyr SM/CEN/SCJW/BI-01-C

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55921#issuecomment-1473699190 after 12:38:07.

codecoup-tester commented:
Failed to merge the branch

sjanc commented:
#AutoPTS run zephyr SM/CEN/SCJW/BI-01-C

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55921#issuecomment-1475795810 after 09:20:21.

codecoup-tester commented:
AutoPTS Bot results:
No failed test found.

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -824,13 +824,13 @@ int ull_conn_rx(memq_link_t *link, struct node_rx_pdu **rx)
 	switch (pdu_rx->ll_id) {
 	case PDU_DATA_LLID_CTRL:
 	{
-		ARG_UNUSED(link);
```
not really related to this PR, but the next line should be removed as well

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -279,6 +279,38 @@ void llcp_tx_resume_data(struct ll_conn *conn, enum llcp_tx_q_pause_data_mask re
+	ctx->node_ref.rx->hdr.link = ctx->node_ref.link;
+}
+
+void llcp_nodes_release(struct ll_conn *conn, struct proc_ctx *ctx)
```
is this functionality tested in unit or babblesim tests, for example tests/bsim/bluetooth/ll/conn/tests_scripts/basic_conn20_split.sh?

erbr-ot replied:
It is tested in the basic_conn20_split - in the sense that I saw the test failing (due to inability to alloc RX nodes in the scanner) without the release flow. And I noticed - by inspecting printk's during dev that alloc'ed RX nodes were correctly released after applying the change

kruithofa commented:
@erbr-ot some minor things, but nothing that should block

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -279,6 +279,38 @@ void llcp_tx_resume_data(struct ll_conn *conn, enum llcp_tx_q_pause_data_mask re
+		conn->llcp.rx_node_release = ctx->node_ref.rx;
+	}
+#if defined(CONFIG_BT_CTLR_PHY) && defined(CONFIG_BT_CTLR_DATA_LENGTH)
+	if (ctx->proc == PROC_PHY_UPDATE && ctx->data.pu.ntf_dle_node) {
```
No, since they can be - if we have both a pending dle and phy ntf, they must both be released

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -1124,40 +1125,31 @@ static void rp_comm_st_idle(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
 	ARG_UNUSED(pdu);
 	/* Allocate ntf node */
-	ntf = llcp_ntf_alloc();
+	ntf = ctx->node_ref.rx;
```
ARG_UNUSED removed

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_local.c:
```diff
@@ -171,7 +171,7 @@ struct proc_ctx *llcp_lr_peek(struct ll_conn *conn)
 bool llcp_lr_ispaused(struct ll_conn *conn)
 {
-	return conn->llcp.local.pause == 1U;
+	return (conn->llcp.local.pause == 1U);
```
braces removed

erbr-ot commented on tests/bluetooth/controller/ctrl_phy_update/src/main.c:
```diff
@@ -120,7 +121,7 @@ ZTEST(phy_central, test_phy_update_central_loc)
 	struct pdu_data_llctrl_phy_req rsp = { .rx_phys = PHY_1M | PHY_2M,
 					       .tx_phys = PHY_1M | PHY_2M };
-	struct pdu_data_llctrl_phy_upd_ind ind = { .instant = 7,
+	struct pdu_data_llctrl_phy_upd_ind ind = { .instant = 8,
```
Well, indirectly as the TX of UPD_IND is held off one conn event due to missing TX node for NTF, so one extra event passes before the UPD_IND is sent, thus instant +=1

erbr-ot commented on tests/bluetooth/controller/ctrl_phy_update/src/main.c:
```diff
@@ -1128,13 +1183,19 @@ ZTEST(phy_periph, test_phy_update_periph_loc_collision)
 	/* Done */
 	event_done(&conn);
 
+	/* Prepare */
```
Yes - the WAIT_FOR_INSTANT_ON_AIR - and I figured the easiest was to simple allow another event to pass in the test

kruithofa reviewed (dismissed):
LGTM

erbr-ot committed:
Bluetooth: test: fixing callback declaration in basic conn bsim test

Using the macro for declaring the callback struct does not make
sense in this construct, where test_connect1/2 are compiled into one
image and specific test functionality is selected via test ID

This reverts change made by #8e1682d which is incorrect for the cases
where the compile happens to include more than one declaration of the
callback struct.

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
Bluetooth: test: adding bsim basic conn test of multiple re-connects

Adding a test that confirms ability to perform 20 (re)connections
between simple peripheral and central

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot commented on tests/bsim/bluetooth/ll/conn/src/test_connect1.c:
Very good point! Thanks.

erbr-ot replied:
Commit message updated and author of https://github.com/zephyrproject-rtos/zephyr/commit/8e1682d1ea5865fd1a38f22546d2eeb7cb026f53 notified

cvinayak reviewed (approved):
quick-ack. Pending, CI testing of multiple CIS in a draft PR here: https://github.com/zephyrproject-rtos/zephyr/pull/56681
If something fails, bugs be fixed in follow up PR.

cvinayak reviewed (changes requested):
@erbr-ot  BabbleSim test that establishes 9 CISes (ACL then CIS on that ACL, then next ACL then CIS on that ACL...) fails with the changes in this PR, see https://github.com/zephyrproject-rtos/zephyr/pull/56681

erbr-ot committed:
Bluetooth: test: removing nrf51dk_nrf51422 from init test target list

Due to increase in image size the target nrf51dk_nrf51422 no longer
supports this test.

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

cvinayak commented:
Please rebase, new babblesim tests now upstream.

erbr-ot commented:
Please not that no changes were made to this apart from rebases since approvals in march

cvinayak reviewed (dismissed):
quick-ack, as most code was reviewed in a prior PR that was reverted in place of this new PR.

kruithofa reviewed (dismissed):
LGTM
